stdlib/universe/mode.go

```diff
@@ -0,0 +1,643 @@
+func NewModeTransformation(d execute.Dataset, cache execute.TableBuilderCache, spec *ModeProcedureSpec) *modeTransformation {
+	return &modeTransformation{
+		d:      d,
+		cache:  cache,
+		column: spec.Column,
+	}
+}
+
+func (t *modeTransformation) RetractTable(id execute.DatasetID, key flux.GroupKey) error {
+	return t.d.RetractTable(key)
+}
+
+func (t *modeTransformation) Process(id execute.DatasetID, tbl flux.Table) error {
+	builder, created := t.cache.TableBuilder(tbl.Key())
+	if !created {
+		return fmt.Errorf("mode found duplicate table with key: %v", tbl.Key())
+	}
+
+	colIdx := execute.ColIdx(t.column, tbl.Cols())
+	if colIdx < 0 {
+		// doesn't exist in this table, so add an empty value
+		if err := execute.AddTableKeyCols(tbl.Key(), builder); err != nil {
+			return err
+		}
+		colIdx, err := builder.AddCol(flux.ColMeta{
+			Label: execute.DefaultValueColLabel,
+			Type:  flux.TString,
+		})
+		if err != nil {
+			return err
+		}
+
+		if err := builder.AppendString(colIdx, ""); err != nil {
+			return err
+		}
+		if err := execute.AppendKeyValues(tbl.Key(), builder); err != nil {
+			return err
+		}
+		// TODO: hack required to ensure data flows downstream
+		return tbl.Do(func(flux.ColReader) error {
+			return nil
+		})
+	}
+
+	col := tbl.Cols()[colIdx]
+
+	if err := execute.AddTableKeyCols(tbl.Key(), builder); err != nil {
+		return err
+	}
+
+	oldIdx := colIdx
+
+	colIdx, err := builder.AddCol(flux.ColMeta{
+		Label: execute.DefaultValueColLabel,
+		Type:  col.Type,
+	})
+	if err != nil {
+		return err
+	}
+
+	if tbl.Key().HasCol(t.column) {
+		j := execute.ColIdx(t.column, tbl.Key().Cols())
+
+		if err := builder.AppendValue(colIdx, tbl.Key().Value(j)); err != nil {
+			return err
+		}
+
+		if err := execute.AppendKeyValues(tbl.Key(), builder); err != nil {
+			return err
+		}
+		// TODO: hack required to ensure data flows downstream
+		return tbl.Do(func(flux.ColReader) error {
+			return nil
+		})
+	}
+
+	switch col.Type {
+	case flux.TBool:
+		return tbl.Do(func(cr flux.ColReader) error {
+			return t.doBool(cr, tbl, builder, oldIdx, colIdx)
+		})
+	case flux.TInt:
+		return tbl.Do(func(cr flux.ColReader) error {
+			return t.doInt(cr, tbl, builder, oldIdx, colIdx)
+		})
+	case flux.TUInt:
+		return tbl.Do(func(cr flux.ColReader) error {
+			return t.doUInt(cr, tbl, builder, oldIdx, colIdx)
+		})
+	case flux.TFloat:
+		return tbl.Do(func(cr flux.ColReader) error {
+			return t.doFloat(cr, tbl, builder, oldIdx, colIdx)
+		})
+	case flux.TString:
+		return tbl.Do(func(cr flux.ColReader) error {
+			return t.doString(cr, tbl, builder, oldIdx, colIdx)
+		})
+	case flux.TTime:
+		return tbl.Do(func(cr flux.ColReader) error {
+			return t.doTime(cr, tbl, builder, oldIdx, colIdx)
+		})
+	}
+
+	if err := execute.AppendKeyValues(tbl.Key(), builder); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (t *modeTransformation) doString(cr flux.ColReader, tbl flux.Table, builder execute.TableBuilder, oldIdx int, colIdx int) error {
```

Comment:
In Go, if you have repeated arguments with the same type, you can do `oldIdx, colIdx int`. Can you also rename these to be `srcColIdx` and `destColIdx`?